### PR TITLE
New gatsby-remark-embed-snippet plug-in

### PR DESCRIPTION
**Pull request by bvaughn**
For a more detailed explanation of this plug-in and its options, refer to the [README](https://github.com/bvaughn/gatsby/blob/3ae567a0380047637a93de294c276faed33a79d4/packages/gatsby-remark-embed-snippet/README.md).

Given a file like this (eg `examples/hello-world.js`):
```js
// highlight-range{1,4}
ReactDOM.render(
  <h1>Hello, world!</h1>,
  document.getElementById('root')
);
```

And a markdown snippet like this:
```md
Here is some embedded code:
`embed:hello-world.js`
```

Create HTML like this:
screen shot 2017-11-23 at 4 26 18 pm

This PR is related to ReactJS.org issue reactjs/reactjs.org/issues/310 and PR reactjs/reactjs.org/pull/331

## Follow-up Work

I think the highlight marker comments from this PR would be a nice addition to the `gatsby-remark-prismjs` plugin as well. Defining ranges at the top of the markdown comment is kind of painful, especially for larger blocks or edits. Ideally this plug-in could be simplified to just read files from disk and pass their contents along to a method shared with that plug-in to convert to Prism blocks.

I'd be willing to make this refactor as a follow-up PR if you'd be interested?

### CODE REVIEW

**Comment by bvaughn** [comment]
I kind of hate that "Strings must use backtick quotes" lint rule 😁

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit bded111920d6d4a04adb0b39b716d3019aed9f77

https://app.netlify.com/sites/using-glamor/deploys/5a17652ba1147749ece5b65f

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 08f222e5e0368a172d6900a259c52f5f54fcd52d

https://app.netlify.com/sites/using-glamor/deploys/5a1784684c4b937f4f0b3164

**Comment by bvaughn** [comment]
I'm going to add range support here momentarily.

**Comment by bvaughn** [comment]
I think the highlight marker comments from this PR could be really useful for the `gatsby-remark-prismjs` plugin as well. Ideally, this plug-in could be simplified to just read files from disk and then pass the contents along to the same util methods used by the other plug-in to convert to Prism blocks.

I'd be willing to make this refactor as a follow-up PR if you'd be interested?

**Comment by bvaughn** [comment]
I left a note about possible follow-up work on the PR. If you're interested in merging that, let me know and I'll be happy to submit another PR.

If you're not interested, that's okay too. Let me know and I'll just roll my own plug-in. 😄

**Comment by KyleAMathews** [comment]
>I kind of hate that "Strings must use backtick quotes" lint rule

This is the best rule ever :-D

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
+"use strict"
+
+const fs = require(`fs`)
+const normalizePath = require(`normalize-path`)
+const rangeParser = require(`parse-numeric-range`)
+const visit = require(`unist-util-visit`)
+
+// HACK: It would be nice to find a better way to share this utility code.
+const highlightCode = require(`gatsby-remark-prismjs/highlight-code`)
```
Comment (KyleAMathews) [inline packages/gatsby-remark-embed-snippet/src/index.js]: What we've been doing is to just split it out into a new plugin.

**Comment by KyleAMathews** [comment]
>I think the highlight marker comments from this PR would be a nice addition to the gatsby-remark-prismjs plugin as well. Defining ranges at the top of the markdown comment is kind of painful, especially for larger blocks or edits. Ideally this plug-in could be simplified to just read files from disk and pass their contents along to a method shared with that plug-in to convert to Prism blocks.

100% agree! It's super obnoxious haha :-) Inline highlight comments are way better. And yeah, allowing code re-use is always very nice too.

If this is ready to go, I'll merge/publish and wait looking forward to your follow-up PR :-)

**Comment by bvaughn** [comment]
> If this is ready to go, I'll merge/publish and wait looking forward to your follow-up PR :-)

This should be ready to go, yes. I've got a branch of the reactjs.org docs built using this plug-in just to test it out. 😄 

I'll send a follow-up PR sometime this week.

**Comment by KyleAMathews** [comment]
`spyOn` is a Jest global? It's breaking lint test — so should be added to our .eslintrc globals.

**Comment by bvaughn** [comment]
> `spyOn` is a Jest global? It's breaking lint test — so should be added to our .eslintrc globals.

Yup it's a global. It's also used [here](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-remark-code-repls/src/__tests__/gatsby-node.js#L73) (although that's my fault too I guess 😁 )

**Comment by KyleAMathews** [comment]
Huh, weird the other place doesn't fail the lint too.

**Comment by bvaughn** [comment]
> Huh, weird the other place doesn't fail the lint too.

Agreed!

I'll add an entry to the ESLint config though.

**Comment by bvaughn** [comment]
I believe it should be good to go now. `yarn lint` passes locally at least.

**Comment by KyleAMathews** [comment]
Cool! This is awesome stuff, looking forward to seeing it rolled out on reactjs.org!

**Pull request merged**